### DISCUSSION TITLE
How to load ResourceBundles outside classpath?

### DISCUSSION THREAD

**Discussion by ls-bit** [oracle/graal#8000] [Q&A] [answered]
We are trying to load resource bundles from arbitrary paths, as we need to load some files dynamically at runtime. Following example creates an URLClassLoader from a file and loads a ResourceBundle. It works fine when running as jar:
```java
// from https://stackoverflow.com/questions/1172424/how-to-load-a-resource-bundle-from-a-file-resource-in-java
File file = new File(".");
URL[] urls = { file.toURI().toURL() };
ClassLoader loader = new URLClassLoader(urls);
ResourceBundle rb = ResourceBundle.getBundle("messages", Locale.ENGLISH, loader);
```
When running as native image it only works when files are on available on classpath and configured according to https://www.graalvm.org/jdk21/reference-manual/native-image/dynamic-features/Resources/.

However, it fails when we try to load ResourceBundles from paths outside classpath with MissingResourceException. We tried adding `-H:-LocalizationOptimizedMode`, but no difference.

Is loading ResourceBundles outside classpath supported for native image or is it not possible?

**Comment by alina-yur** [comment]
Hi @ls-bit, I'm afraid loading ResourceBundles outside classpath won’t work with Native Image.

**Comment by vjovanov** [comment, marked as answer]
We plan to remove the flag `LocalizationOptimizedMode`. Once this is done, we should be able to support this case.

**Reply by ls-bit** [reply]
Thanks for the replies! As a workaround we are loading property based ResourceBundles from filesystem using a custom ResourceBundle.Control https://stackoverflow.com/a/6819440

**Comment by loicottet** [comment]
This use case is now supported in Native Image. Since the resource bundle lookup can perform class and resource queries, it requires the bundle name to be registered in a `META-INF/native-image/reachability-metadata.json` file with the following syntax:
```json
{
  "resources": [
    {
      "bundle": "messages"
    }
  ]
}
```
